### PR TITLE
Simplify instantiating SyntheticFieldLoader instances

### PR DESCRIPTION
**Pull request by martijnvg**
- Exclude `SourceLoader.SyntheticFieldLoader.NOTHING` loaders. These have nothing to add. Most meta field mappers return this instance. Only doc count field mapper returns a real synthetic loader.
- Only concat streams in root object mapper and not for any other object mapper.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 1]
```diff
@@ -712,19 +712,17 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
...
-            Stream.concat(extra, mappers.values().stream())
-                .sorted(Comparator.comparing(Mapper::name))
-                .map(Mapper::syntheticFieldLoader)
-                .filter(l -> l != null)
-                .toList()
-        );
+    public SourceLoader.SyntheticFieldLoader syntheticFieldLoader(Stream<Mapper> mappers) {
+        var fields = mappers.sorted(Comparator.comparing(Mapper::name))
+            .map(Mapper::syntheticFieldLoader)
+            .filter(l -> l != SourceLoader.SyntheticFieldLoader.NOTHING)
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:718]: No `SyntheticFieldLoader` returns `null`, so instead we should exclude the ones that return `NOTHING`. Most meta field mappers return `NOTHING`, except `DocCountFieldMapper`.

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 2]
```diff
@@ -712,19 +712,17 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
...
+            .toList();
+        return new SyntheticSourceFieldLoader(fields);
     }
 
     @Override
     public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
-        return syntheticFieldLoader(Stream.empty());
+        return syntheticFieldLoader(mappers.values().stream());
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:725]: Why is this needed?

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:725]: The `syntheticFieldLoader(...)` method now only loads `SyntheticFieldLoader` instances for the fields that are provided. This allows us to only stream concat in the place we need to (root level).

**Review by kkrik-es — APPROVED** [review]
